src/datreant/core/metadata.py

```diff
@@ -1,63 +1,512 @@
 """
-AggLimbs are limbs specifically built for collections, in particular
-:class:`datreant.core.collections.Bundle`. They often correspond to Treant
-limbs but serve as aggregators over collections of them.
-
+Metadata components (tags, categories) give texture to directories, allowing
+for filtering and selection.
 """
+import os
 import itertools
 import functools
-from six import string_types, with_metaclass
+from six import string_types
+from six.moves import range
+from collections import defaultdict
 
 from fuzzywuzzy import process
 
-from . import filesystem
-from . import _AGGTREELIMBS, _AGGLIMBS
-from .collections import Bundle
-from .limbs import Tags, Categories
+from .state import JSONFile
+from .selectionparser import parse_selection
 
 
-class _AggTreeLimbmeta(type):
-    def __init__(cls, name, bases, classdict):
-        type.__init__(type, name, bases, classdict)
+class Metadata(object):
 
-        limbname = classdict['_name']
-        _AGGTREELIMBS[limbname] = cls
+    # TODO; raise meaningful exceptions when file can't be read, written, read
+    # and nonexistent, missing, or treantdir missing
 
+    def __init__(self, tree):
+        self._tree = tree
 
-class _AggLimbmeta(type):
-    def __init__(cls, name, bases, classdict):
-        type.__init__(type, name, bases, classdict)
+    @property
+    def _write(self):
+        self._statefile = JSONFile(os.path.join(self._tree._treantdir,
+                                                self._statefilename),
+                                   init_state=self._init_state)
 
-        limbname = classdict['_name']
-        _AGGLIMBS[limbname] = cls
+        return self._statefile.write()
 
+    @property
+    def _read(self):
+        self._statefile = JSONFile(os.path.join(self._tree._treantdir,
+                                                self._statefilename),
+                                   init_state=self._init_state)
 
-class AggTreeLimb(with_metaclass(_AggTreeLimbmeta, object)):
-    """Core functionality for limbs attached to a View.
+        return self._statefile.read()
+
+
+@functools.total_ordering
+class Tags(Metadata):
+    """Interface to tags.
 
     """
-    _name = 'aggtreelimb'
+    _statefilename = 'tags.json'
 
-    def __init__(self, collection):
-        self._collection = collection
+    @staticmethod
+    def _init_state(jsonfile):
```

Comment:
why is this function really needed? I don't see the benefit of it.

Reply:
I'm also confused here why this function is static. It is only ever called with `self` as argument.

Reply:
This method is the `init_state` method specific to `Tags`, which gets fed to the `JSONFile` constructor for use when creating a file from scratch. For `Tags` this generates an empty list. `Categories` has its own that generates an empty dict.